### PR TITLE
Replace Where Grad "Mul" with "Where"

### PR DESCRIPTION
**Pull request by harshithapv**
Replace Where Grad "Mul" with "Where". This removes NaNs with fp32 runs for XSoftmax's -inf mask in HF-Deberta. 

image

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -1345,16 +1345,16 @@ IMPLEMENT_GRADIENT_BUILDER(GetSliceGradient) {
...
 IMPLEMENT_GRADIENT_BUILDER(GetWhereGradient) {
   std::vector<NodeDef> result;
-  const int64_t data_type = static_cast<int64_t>(IElemType(1));
+  NodeDef zero_constant_node = ZeroConstantNode(OElemType(0));
+  ArgDef ZERO = zero_constant_node.output_args[0];
+  result.push_back(zero_constant_node);
   if (IsGradientRequiredForSrcNodeInput(1)) {
-    result.push_back(NodeDef("Cast", {I(0)}, {IA("Positive_Mask")}, {MakeAttribute("to", data_type)}));
-    result.push_back(NodeDef("Mul", {GO(0), IA("Positive_Mask")}, {GI(1)}));
+    result.push_back(NodeDef("Where", {I(0),GO(0),ZERO}, {GI(1)}));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1352]: please turn on auto format....

**Review by SherlockNoMad — DISMISSED** [review]

**Review by SherlockNoMad — APPROVED** [review]

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 2]
```diff
@@ -1345,16 +1345,16 @@ IMPLEMENT_GRADIENT_BUILDER(GetSliceGradient) {
...
+  ArgDef ZERO = zero_constant_node.output_args[0];
+  result.push_back(zero_constant_node);
   if (IsGradientRequiredForSrcNodeInput(1)) {
-    result.push_back(NodeDef("Cast", {I(0)}, {IA("Positive_Mask")}, {MakeAttribute("to", data_type)}));
-    result.push_back(NodeDef("Mul", {GO(0), IA("Positive_Mask")}, {GI(1)}));
+    result.push_back(NodeDef("Where", {I(0), GO(0), ZERO}, {GI(1)}));
   }
 
   if (IsGradientRequiredForSrcNodeInput(2)) {
     result.push_back(NodeDef("Not", {I(0)}, {IA("Not_Condition", IType(0))}));
```
Comment (ashbhandare) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1356]: You can just flip the input 2 and 3 of where instead of adding a Not

Comment (harshithapv) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1356]: Ah yes. I will do the change.

Comment (harshithapv) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: tested this with HF-Deberta. No convergence or perf issue seen.

**Review by ashbhandare — APPROVED** [review]

**Pull request merged**